crates/bevy_render/src/pipeline_keys/mod.rs

```diff
@@ -0,0 +1,682 @@
+    pub fn get_key<'a, K: PipelineKeyType>(
+        &self,
+        cache: &'a PipelineCache,
+    ) -> Option<PipelineKey<'a, K>> {
+        Some(cache.key_store().pipeline_key(self.get_raw::<K>()?))
+    }
+}
+
+/// A wrapper around the underlying `PipelineKeyType` that also contains a reference to the `KeyMetaStore`,
+/// allowing access to dynamic key components, and to `ShaderDefVal`s that the key's value (or components) specify.
+#[derive(Debug, Copy, Clone)]
+pub struct PipelineKey<'a, T: PipelineKeyType> {
+    store: &'a KeyMetaStore,
+    value: T,
+}
+
+impl<'a, T: PipelineKeyType> Deref for PipelineKey<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+impl<'a, T: PipelineKeyType + PartialEq> PartialEq for PipelineKey<'a, T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl<'a, T: PipelineKeyType> PipelineKey<'a, T> {
+    /// Extract a `PipelineKey` for a given component of this key, if it is present.
+    /// This is mainly useful for dynamic keys (on fixed keys we can just access the members directly if required),
+    /// but may also be useful for obtaining a `PipelineKey` wrapper of a component.
+    pub fn try_extract<U: PipelineKeyType>(&'a self) -> Option<PipelineKey<'a, U>> {
+        let positions = T::positions(self.store);
+        let SizeOffset { size, offset } = positions.get(&TypeId::of::<U>())?;
+        let key = T::pack(&self.value, self.store);
+        let value = (key.packed >> offset) & ((1 << size) - 1);
+        Some(self.store.pipeline_key(value))
+    }
+
+    /// Extract a `PipelineKey` for a given component of this key, panics if not found.
+    pub fn extract<U: PipelineKeyType>(&'a self) -> PipelineKey<'a, U> {
+        let positions = T::positions(self.store);
+        let SizeOffset { size, offset } = positions
+            .get(&TypeId::of::<U>())
+            .unwrap_or_else(missing::<U, _>);
+        let key = T::pack(&self.value, self.store);
+        let value = (key.packed >> offset) & ((1 << size) - 1);
+        self.store.pipeline_key(value)
+    }
+
+    /// construct a `PipelineKey` for an arbitrary type, using the `KeyMetaStore` reference from this key.
+    pub fn construct<K: PipelineKeyType>(&'a self, value: K) -> PipelineKey<'a, K> {
+        PipelineKey {
+            store: self.store,
+            value,
+        }
+    }
+
+    /// Get a list of shader defs corresponding to this key's value (and of any components)
+    pub fn shader_defs(&'a self) -> Vec<ShaderDefVal> {
+        debug!(
+            "{} shader defs, positions: {:?}",
+            type_name::<T>(),
+            T::positions(self.store)
+        );
+        T::shader_defs(T::pack(&self.value, self.store).packed, self.store)
+    }
+}
+
+/// Implemented for non-dynamic keys, allows access to the bitsize without providing the `KeyMetaStore`
+pub trait FixedSizeKey: 'static {
+    fn fixed_size() -> u8;
```

Comment:
perhaps ... would still need manual impls for bool, Option<T> and the wgpu types. could omit the impls for u8, u32, i32. 

i think since the default is often wrong it's probably better to require it to be done explicitly.